vm/actor/src/builtin/verifreg/state.rs

```diff
@@ -0,0 +1,153 @@
+        State {
+            root_key,
+            verifiers: empty_map.clone(),
+            verified_clients: empty_map,
+        }
+    }
+
+    pub fn put_verifier<BS: BlockStore>(
+        &mut self,
+        store: &BS,
+        verified_addr: &Address,
+        verifier_cap: &Datacap,
+    ) -> StateResult<()> {
+        self.verifiers = Self::put(&mut self.verifiers, store, verified_addr, verifier_cap)?;
+        Ok(())
+    }
+
+    pub fn get_verifier<BS: BlockStore>(
+        &mut self,
+        store: &BS,
+        address_get: &Address,
+    ) -> StateResult<Option<Datacap>> {
+        Self::get(&mut self.verifiers, store, address_get)
+    }
+
+    pub fn delete_verifier<BS: BlockStore>(
+        &mut self,
+        store: &BS,
+        address: &Address,
+    ) -> StateResult<()> {
+        self.verifiers = Self::delete(&mut self.verifiers, store, address)?;
+        Ok(())
+    }
+
+    pub fn put_verified_client<BS: BlockStore>(
+        &mut self,
+        store: &BS,
+        verified_addr: &Address,
+        verifier_cap: &Datacap,
+    ) -> StateResult<()> {
+        self.verified_clients = Self::put(
+            &mut self.verified_clients,
+            store,
+            verified_addr,
+            verifier_cap,
+        )?;
+
+        Ok(())
+    }
+
+    pub fn get_verified_client<BS: BlockStore>(
+        &mut self,
+        store: &BS,
+        address: &Address,
+    ) -> StateResult<Option<Datacap>> {
+        Self::get(&mut self.verified_clients, store, address)
+    }
+
+    pub fn delete_verified_client<BS: BlockStore>(
+        &mut self,
+        store: &BS,
+        address: &Address,
+    ) -> StateResult<()> {
+        self.verified_clients = Self::delete(&mut self.verified_clients, store, address)?;
+        Ok(())
+    }
+
+    //private helper functions
+    fn put<BS: BlockStore>(
+        storage: &mut Cid,
+        store: &BS,
+        verified_addr: &Address,
+        verifier_cap: &Datacap,
+    ) -> StateResult<Cid> {
+        let mut map: Hamt<BytesKey, _> =
+            Hamt::load_with_bit_width(&storage, store, HAMT_BIT_WIDTH)?;
+        map.set(verified_addr.to_bytes().into(), BigUintSer(&verifier_cap))?;
+        let root = map.flush()?;
+        Ok(root)
+    }
+
+    fn get<BS: BlockStore>(
+        storage: &mut Cid,
+        store: &BS,
+        verified_addr: &Address,
+    ) -> StateResult<Option<Datacap>> {
+        let map: Hamt<BytesKey, _> = Hamt::load_with_bit_width(&storage, store, HAMT_BIT_WIDTH)?;
+        Ok(map
+            .get::<_, BigUintDe>(&verified_addr.to_bytes())?
+            .map(|s| s.0))
+    }
+
+    fn delete<BS: BlockStore>(
+        storage: &mut Cid,
+        store: &BS,
+        verified_addr: &Address,
+    ) -> StateResult<Cid> {
+        let mut map: Hamt<BytesKey, _> =
+            Hamt::load_with_bit_width(&storage, store, HAMT_BIT_WIDTH)?;
+        map.delete(&verified_addr.to_bytes())?;
+        let root = map.flush()?;
+        Ok(root)
```

Comment:
Shouldn't need to pass the root back if you are exposing a mutable reference to the storage cid in this function, could just do something like:
```suggestion
        *storage = map.flush()?;
        Ok(())
```

Or am I missing something? if not then you can at least change storage to be immutable ref

Reply:
This is essentially the part with the helper function. The helper function for the verification and verification client. Returning the root just adds a contract with a caller

Reply:
The flushing is happening in the helper function.  for the put and delete. 

Forgot to remove the mut references

Reply:
I guess the HAMT functions already put some side effects in so it is a debate of whether just because we have some side effects, should we put some more basically eitherway fine with any method

Reply:
side effects you are referring to writing to blockstore on flush? Doing this only modifies the cid, and it's being changed still just after the cid is returned. And to clarify this change does not invalidate the previous cid in any way.

I'm good with this now, just curious the thought process behind this (interesting to get the functional perspective to this)

Reply:
Well from a pure functional perspective we wouldn't even be making changes to the HAMT,we would essentially be building a new HAMT strucutre with every operation that we do. A bit overhead here but if we had to take not so much of a deep step functionality, I would rather reduce the amount of mut parameters that these function takes and isolate them a bit at a higher level. The general idea is that if you give these functions certain number of parameters, you know what you are going to get out i mean if the HAMT was a functional data structure this would be more true but the HAMT's version of mutability is a bit more isolated which I think it is fine

Reply:
That is what's happening, the Hamt is being built for every operation (it's very cheap to load because you are just pulling the root node into memory).

That makes sense though to have the mutability further up